lib: fix inspector links for stack traces

Dragiyski opened the pull request:
When the chrome inspector (chrome://inspect) display stack traces, it
makes the filename clickable, and developer can easily locate the
relevant source with a single click. This only works when filenames are
valid URLs. The CJS loader seems to compile sources with absolute path
instead of URL, breaking the inspector functionality. This is especially
apparent when node_modules contain "at" symbol (@). The loader already
presents module to the inspector using file URL, the same URL should be
used for compiling the source.

For example the current stack trace looks like this:

![Current Stack Trace](https://i.imgur.com/D4cj2QJ.png)
```
TypeError: (...)
    at Object.<anonymous> (/home/dragiyski/(...)/@dragiyski/(...)/test.js:19:5)
```
and as shown the link covers only the part after the first slash after the ``@`` directory. The link leads to URL starting with slash (/), which opens an empty (about:blank) tab within the browser. This is an expected behavior, because in URLs ``@`` represents ``user@host`` delimiter and the path is not valid absolute URL as it does not have protocol. A properly fixed URL looks like this:

![Fixed Stack Trace](https://i.imgur.com/3SOLHOW.png)
```
TypeError: (...)
    at Object.<anonymous> (file:///home/dragiyski/(...)/@dragiyski/(...)/test.js:19:5)
```
and clicking on the link provided in inspector opens the actual location of the source for all call sites, including node_modules containing ``@`` symbol in their path. For the console, the path is printed with ``file://`` in front and clickable in some terminals.

Note: (...) replaces a valid path omitted for security reasons.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Dragiyski committed:
fix: inspector links for stack traces

When the chrome inspector (chrome://inspect) display stack traces, it
makes the filename clickable, and developer can easily locate the
relevant source with a single click. This only works when filenames are
valid URLs. The CJS loader seems to compile sources with absolute path
instead of URL, breaking the inspector functionality. This is especially
apparent when node_modules contain "at" symbol (@). The loader already
presents module to the inspector using file URL, the same URL should be
used for compiling the source.

Dragiyski committed:
fix: assert using file urls

It seems getErrMessage() function uses filename from a CallSite to
open a file and extract source code location of the statement. The
CallSite now returns file URL, so it must be converted back to path.

Dragiyski committed:
fix: test using file URL

Various tests explore the stack trace or CallSite objects' filename.
They are modified now to match file URL instead.

ljharb commented on lib/internal/source_map/source_map_cache.js:
```diff
@@ -171,7 +171,16 @@ function sourcesToAbsolute(base, data) {
       source = resolve(base, source);
     }
-    if (!source.startsWith('file://')) source = `file://${source}`;
+    if (!source.startsWith('file://')) {
```
```suggestion
    if (!StringPrototypeStartsWith(source, 'file://')) {
```

Dragiyski replied:
This file does not import ``StringPrototypeStartsWith``. I have added it from ``primordials`` and replaced it in one more location.

ljharb replied:
Thanks, much appreciated :-)

Trott commented on test/sequential/test-cli-syntax-bad.js:
```diff
@@ -42,7 +43,8 @@ const syntaxErrorRE = /^SyntaxError: \b/m;
       // stderr should include the filename
-      assert(stderr.startsWith(file), `${stderr} starts with ${file}`);
+      assert(stderr.startsWith(pathToFileURL(file)),
+             `${stderr} starts with ${file}`);
```
```suggestion
             `${stderr} starts with ${pathToFileURL(file)}`);
```

Trott commented:
This could use some reviews. @nodejs/inspector @nodejs/assert @nodejs/modules-active-members @nodejs/testing @bcoe (source maps)

I keep meaning to take a closer look and don't, so I'll just lazily ask my question here: How does this affect non-inspector output? Will file URLs start showing up in my stack traces at the command line? Or is this inspector-output only?

Dragiyski commented:
> This could use some reviews. @nodejs/inspector @nodejs/assert @nodejs/modules-active-members @nodejs/testing @bcoe (source maps)
> 
> I keep meaning to take a closer look and don't, so I'll ask my question here without having done that (sorry): How does this affect non-inspector output? Will file URLs start showing up in my stack traces at the command line? Or is this inspector-output only?

Yes, it will affect the console output (stderr/stdout) and the inspector output. Current implementation seems inconsistent, there are already URLs in the output when ESM modules are involved:

```javascript
// error.mjs
export default function test() {
    throw new Error();
}
```

```javascript
// main.js
(function () {
    'use strict';

    (async () => {
        const test = await import('./error.mjs');
        test.default();
    })().catch(error => {
        console.error(error);
    });
})();
```

```
Output:
Error
    at Module.test (file:///home/dragiyski/projects/nodejs/test/error.mjs:2:11)
    at /home/dragiyski/projects/nodejs/test/main.js:6:21
```

There is also URLs even the error is thrown in the CJS module:

```javascript
// main.mjs
import test from './error.js';

test();
```

```javascript
// error.js
module.exports = function () {
    throw new Error();
};
```

```
Output:
/home/dragiyski/projects/nodejs/test/error.js:2
    throw new Error();
          ^

Error
    at module.exports (/home/dragiyski/projects/nodejs/test/error.js:2:11)
    at file:///home/dragiyski/projects/nodejs/test/main.mjs:3:1
    at ModuleJob.run (node:internal/modules/esm/module_job:146:23)
    at async Loader.import (node:internal/modules/esm/loader:165:24)
    at async Object.loadESM (node:internal/process/esm_loader:68:5)
```

As it can be seen, ESM modules already report local files as URLs. If any module on NPM examine ``CallSite`` object to retrieve a filename and use that filename to ``fs`` functions, it would break if it was included from ESM module. This will make the output consistent, so any new module would not need to check if ``CallSite`` is an URL (ESM, native node modules, etc) or absolute path (CJS), it would always be an URL.

P.S. It might be useful to allow using ``file:`` URLs in ``fs``.

Trott commented:
It looks like this needs some changes for tests to pass on Windows. 

```
06:53:28 not ok 512 parallel/test-source-map-enable
06:53:28   ---
06:53:28   duration_ms: 0.664
06:53:28   severity: fail
06:53:28   exitcode: 1
06:53:28   stack: |-
06:53:28     node:assert:108
06:53:28       throw new AssertionError(obj);
06:53:28       ^
06:53:28     
06:53:28     AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
06:53:29     + actual - expected
06:53:29     
06:53:29     + 'file://C:\\workspace\\node-test-binary-windows-js-suites\\node\\test\\fixtures\\source-map'
06:53:29     - 'file:///C:/workspace/node-test-binary-windows-js-suites/node/test/fixtures/source-map'
06:53:29         at Object.<anonymous> (file:///C:/workspace/node-test-binary-windows-js-suites/node/test/parallel/test-source-map-enable.js:90:10)
06:53:29         at Module._compile (node:internal/modules/cjs/loader:1087:30)
06:53:29         at Object.Module._extensions..js (node:internal/modules/cjs/loader:1116:10)
06:53:29         at Module.load (node:internal/modules/cjs/loader:948:32)
06:53:29         at Function.Module._load (node:internal/modules/cjs/loader:789:14)
06:53:29         at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:72:12)
06:53:29         at node:internal/main/run_main_module:17:47 {
06:53:29       generatedMessage: true,
06:53:29       code: 'ERR_ASSERTION',
06:53:29       actual: 'file://C:\\workspace\\node-test-binary-windows-js-suites\\node\\test\\fixtures\\source-map',
06:53:29       expected: 'file:///C:/workspace/node-test-binary-windows-js-suites/node/test/fixtures/source-map',
06:53:29       operator: 'strictEqual'
06:53:29     }
06:53:29   ...
```

Not entirely sure why those are showing up in Jenkins but not in GitHub Actions CI, but I think it's because we have to skip a bunch of `fs` tests on GitHub Actions CI on Windows because of something about the underlying file system that is escaping my memory right now.

richardlau commented:
> Not entirely sure why those are showing up in Jenkins but not in GitHub Actions CI, but I think it's because we have to skip a bunch of `fs` tests on GitHub Actions CI on Windows because of something about the underlying file system that is escaping my memory right now.

We don’t currently run any tests on Windows in GitHub actions. We used to in the workflow for testing building from the source tarball until https://github.com/nodejs/node/pull/34440.

Dragiyski committed:
test: fix test message

bcoe commented:
I'm excited to have another person looking at stack traces 😄

_A couple initial questions (I'll give a more thorough review in the next couple days)._

1. I see you're updating the stack traces when source maps are enabled too (thanks for catching this). Does the inspector handle the linking appropriately for the original call sites added by --enable-source-maps?
2. I think we should take the time to make sure the paths are handled appropriately in the inspector UI on windows as well.

I will be curious to hear other people's feedback. The addition of the scheme seems a little redundant for the terminal output, but I appreciate that it makes for a better experience in the inspector.

Dragiyski committed:
test: use URLs when examine the stack trace

Windows path (using backslash) won't be found in stack trace anymore.
Use appropriate url.fileURLToPath and url.pathToFileURL when examine
the stack trace.

Dragiyski commented:
> 1\. I see you're updating the stack traces when source maps are enabled too (thanks for catching this). Does the inspector handle the linking appropriately for the original call sites added by --enable-source-maps?

Yes, when source maps are enabled, the line plugged in into the stack is clickable and links to the source mapped by the source map. The original line is also shown and it leads to the minified file.

Chrome's devtools client have some kind of link detection algorithm. You can put in the console any link (including ``http://`` or ``https://``) and it will be clickable even when the link is suffixed with colon (:) and a line number. If it matches registered script exactly it will jump to it, otherwise it will open a new tab in the last focused browser window. However this does not properly detects all URLs in general. For node modules ``node:internal/url.js`` will have only ``/url.js`` clickable. However, when printing a stack trace the entire part ``node:internal/url.js`` would be detected properly and it will be clickable. I have no idea why there would be any difference here. Even worse, when not all call sites are detected as URLs, call sites for ``node:`` are shown wrong (URL starts after the first slash), but when all call sites (that have filename) are URLs it works... :confused: It seems having slashes (like ``node://``) helps to detect URL in general, but it does not lead to any registered files. I think that's a bug the devtools devs must fix, but they probably won't because, in websites, all call sites that have filename (they are not native C++) are URLs (scripts in HTML page or remote scripts).

Anyway, the idea here is to register all scripts in ``v8::ScriptCompiler`` as URL, which is possible since v15.x. Node modules in lib are valid URLs with ``node:`` protocol. ESM modules are valid URLs with ``file://`` protocol, even when they are loaded from the local filesystem. The only modules that fail to follow that format are the CJS modules. Because of the ESM modules all facilities to handle path to file URL conversion (or vice versa) are already present and used.

> 2\. I think we should take the time to make sure the paths are handled appropriately in the inspector UI on windows as well.

I tried to address that in ``sourcesToAbsolute``, but there is no safe portable way to manually build a ``file://`` URL using string concatenation and I do not see why we need to do that. The final goal is an URL and ``url`` module already provides suicient utilities for handling that logic even for more complex cases (detecting if URL is already absolute for any protocol). Originally I tried to apply minimal changes to address sources appended to the call stack, but as it is mentioned in #35903, the source could by absolute path from a library.

bcoe commented on lib/assert.js:
```diff
@@ -281,12 +282,18 @@ function getErrMessage(message, fn) {
 
+  if (filename && StringPrototypeStartsWith(filename, 'file:')) {
+    try {
+      filename = fileURLToPath(filename);
```
With the check for `file:`, under what conditions do we expect this to fail?

If we leave this `try/catch`, we might want to put a `debug` in the `catch`.

Dragiyski replied:
``fileURLToPath`` and ``normalizeReferrerURL`` call ``URL`` constructor with a single argument. There was a string (I do not remember exactly, but I can try to remove the try/catch and rebuild to find it) like ``[eval something]`` that makes ``URL`` throws an ``ERR_INVALID_URL`` error. Since that is in the unit tests, it might be used in the future. Moreover, users of ``vm`` can pass any string for filename or have source map containing any string. The combination of both is possible to make otherwise invalid URL to appear as valid module and in general invalid URLs supplied by the user should not crush for the proper reason: ``MODULE_NOT_FOUND`` for example.

bcoe replied:
Might be worth getting a unit test around the edge case if possible 👌

bcoe commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -995,10 +995,14 @@ function wrapSafe(filename, content, cjsModuleInstance) {
   let compiled;
+  let filenameUrl = filename;
+  try {
+    filenameUrl = normalizeReferrerURL(filename);
```
likewise, if we can instead test for filename patterns that would make `normalizeReferrerURL` throw, it might be better than a catch all `try`/`catch`.

bcoe commented on lib/internal/source_map/source_map_cache.js:
```diff
@@ -166,12 +167,16 @@ function sourceMapFromDataUrl(basePath, url) {
+      // If source is any form of full URL (for any protocol)
+      // it will remain full URL. If it is relative, it will
+      // become file:// url, because the baseUrl is file URL.
+      source = (new URL(source, baseUrl)).href;
```
I just ended up undertaking close to this same refactor in https://github.com/nodejs/node/pull/35903

Dragiyski replied:
Yes, that function is affected in both issues. I traced its usage to ensure the first parameter is always a directory path and then converted it to URL. It seems your solution is to first convert it to URL and then call ``sourcesToAbsolute``. Please, make sure it ends with slash ``/``. WHATWG rules for URL will threat the last part as as file otherwise and return one directory upwards. Otherwise both solutions are identical in their core: ``new URL(source, baseURL).href``.

P.S. I am not familiar with the nodejs source, so I try to minimize the amount of dependent changes.

bcoe reviewed (commented):
> It seems having slashes (like node://) helps to detect URL in general, but it does not lead to any registered files. I think that's a bug the devtools devs must fix, but they probably won't because, in websites, all call sites that have filename (they are not native C++) are URLs (scripts in HTML page or remote scripts).

It might be worth talking to some folks who work on inspector, see if they have any thoughts. Where would we want `node:` to link to? _the node.js docs?_, _the source code on GitHub?_

>  tried to address that in sourcesToAbsolute, but there is no safe portable way to manually build a file:// URL using string concatenation and I do not see why we need to do that. 

I'm less concerned about the unit tests passing (I'm sure you'll end up finding the edge cases causing problems).

What I'd like to make sure, is that the inspector UI works as we'd hope on a few testing platforms:

* you've already tested that adding schemes makes the user experience better on OSX/Linux ✅ .
* It would be good to have someone run an inspector session on Windows and confirm the same UX improvement.
* I'd be curious to see how VS Code handles this new style of errors, does it link the error paths too?

Dragiyski commented:
> What I'd like to make sure, is that the inspector UI works as we'd hope on a few testing platforms:

> * It would be good to have someone run an inspector session on Windows and confirm the same UX improvement.

I tested the behavior of devtools on both Windows 10 (Pro June 2020 with Chrome 86.0.4240.111, on VirtualBox) and Ubuntu 20.04, it shows identical behavior:
  * with ``--enable-source-maps``: [Screenshot](https://i.imgur.com/QoHiKIt.png)
    * all ``file://`` URLs are shown in full and clickable. Clicking open the containing file and it works for both the original and the mapped source. The file is scrolled to the specified line and the cursor is set to the column specified in the URL.
    * The ``node:`` URLs for native node modules are URLs after the first slash. Clicking on them does not lead to a source file, but opens ``about:blank`` in the nearest browser.
  * without ``--enable-source-maps``: [Screenshot](https://i.imgur.com/Jf63Qs6.png)
    * All ``file://`` URLs are now reduced to the a base name and still clickable. Even if source maps are not processed by the node, devtools process them and include the original source. In this case there is no link to the mapped source, only to the original specified in the source map. Links scroll to the proper line and put the cursor on the proper column.
    * All ``node:`` are fully shown and this time all the URL is clickable and lead to proper source file, line and column. Internal node modules are listed as ``(no domain)`` as expected and links open the specified file.

With both Webstorm (2020.2.3 on Ubuntu 20.04) ([Screenshot](https://i.imgur.com/99SVVkL.png)) and VS Code (configuration above) ([Screenshot](https://i.imgur.com/jbWXHij.png)) the ``file://`` entire URL is visible and:
  * with ``--enable-source-maps``
    * All ``file://`` URLs are clickable and open the referenced source file, line and column for both the original and mapped file.
    * Other URLs (including ``node:``) are not clickable at all.
  * without ``--enable-source-maps``: same as above, but now since the lines for the original files are missing, there is only link to the mapped file. It does not seems those programs process source maps by themselves by default.

If someone have the tools and the platform, the following is still not tested:
  * Mac OS
  * ``v8::Function`` from a ``v8::FunctionTemplate`` on the stack trace.
  * Negative scenarios:
    * Source maps with invalid URLs, URLs to non-existing original, Invalid source map data, etc.
    * Function compiled by ``vm`` with invalid URLs for a filename.

joyeecheung commented:
module benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/676/

bcoe commented:
I _think_ we should probably call this a SemVer major, as upstream libraries tend to assert against error strings (_which will change to having a `file://` prefix_).

bcoe reviewed (approved):
Thanks for the screen shots, I think this will be a definitely usability improvement for folks using inspectors.

---

Not for this PR, but two questions I'd like to eventually answer:

1. how can we better handle `node://` URIs.
2. how should we handle special paths like `[eval something]`.

bcoe commented:
@nodejs/modules @nodejs/tooling friendly nudge, it would be good to get a few more sets of eyes on this. It will alter stack trace output in a significant way.

Dragiyski commented:
> @nodejs/modules @nodejs/tooling friendly nudge, it would be good to get a few more sets of eyes on this. It will alter stack trace output in a significant way.

Just a reminder that the stack trace is already affected for libraries that use the ``CallSite`` objects. Most notably there is a potential bug in ``assert``. For example:

```javascript
// test-assert.js
const assert = require('assert');

assert(false);
```
produces expected result:
```console
node:assert:388
    throw err;
    ^

AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert(false)

    at Object.<anonymous> (/home/dragiyski/src/node/dragiyski/test-assert.js:3:1)
    at Module._compile (node:internal/modules/cjs/loader:1083:30)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1112:10)
    at Module.load (node:internal/modules/cjs/loader:948:32)
    at Function.Module._load (node:internal/modules/cjs/loader:789:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:72:12)
    at node:internal/main/run_main_module:17:47 {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
```

However, the following:
```javascript
// test-assert.mjs
import assert from 'assert';

assert(false);
```
produces an unexpected result:
```console
node:internal/process/esm_loader:74
    internalBinding('errors').triggerUncaughtException(
                              ^

AssertionError [ERR_ASSERTION]: false == true
    at file:///home/dragiyski/src/node/dragiyski/test-assert.mjs:3:1
    at ModuleJob.run (node:internal/modules/esm/module_job:146:23)
    at async Loader.import (node:internal/modules/esm/loader:165:24)
    at async Object.loadESM (node:internal/process/esm_loader:68:5) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
```
with error produced here:
https://github.com/nodejs/node/blob/8caa6a9d8c153376d86fbed3a58cfae3555ae47b/lib/assert.js#L320
because ``assert`` does not expect file URL from the ``CallSite.getFilename()``.

This is unrelated to this PR and it might need separate issue. Apparently, there is also no unit test to ensure ESM modules works with modules provided by node (but that's okay for now, since they are still experimental). However, for this PR to pass the tests, all affected libraries had to be fixed, since CJS modules are thoroughly tested for the stack trace format.

joyeecheung commented:
From the benchmark CI, this seems to introduce a significant performance regression in the module loader

```
11:28:00                                                                                    confidence improvement accuracy (*)    (**)   (***)
11:28:00  module/module-loader-circular.js n=10000                                                 ***    -14.94 %       ±3.88%  ±5.16%  ±6.72%
11:28:00  module/module-loader-deep.js cache='false' files=1000 ext=''                             ***     -9.33 %       ±3.91%  ±5.20%  ±6.77%
11:28:00  module/module-loader-deep.js cache='false' files=1000 ext='.js'                          ***    -11.86 %       ±3.66%  ±4.87%  ±6.34%
11:28:00  module/module-loader-deep.js cache='true' files=1000 ext=''                                      -4.59 %       ±5.02%  ±6.69%  ±8.74%
11:28:00  module/module-loader-deep.js cache='true' files=1000 ext='.js'                                    1.23 %       ±8.20% ±10.91% ±14.21%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='abs' name='/'                  *     -3.65 %       ±3.55%  ±4.72%  ±6.16%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='abs' name=''                 ***     -6.48 %       ±3.13%  ±4.17%  ±5.45%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='abs' name='/index.js'          *     -2.98 %       ±2.26%  ±3.00%  ±3.91%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='rel' name='/'                 **     -5.11 %       ±3.51%  ±4.68%  ±6.12%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='rel' name=''                         -2.54 %       ±4.92%  ±6.56%  ±8.55%
11:28:00  module/module-loader.js cache='false' n=1000 files=500 dir='rel' name='/index.js'                -2.29 %       ±3.77%  ±5.01%  ±6.52%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='abs' name='/'                   ***    -14.89 %       ±5.16%  ±6.88%  ±8.97%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='abs' name=''                     **     -8.16 %       ±5.35%  ±7.13%  ±9.29%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='abs' name='/index.js'            **     -7.97 %       ±5.01%  ±6.66%  ±8.68%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='rel' name='/'                   ***    -12.30 %       ±6.02%  ±8.04% ±10.52%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='rel' name=''                    ***     -9.43 %       ±4.45%  ±5.92%  ±7.71%
11:28:00  module/module-loader.js cache='false' n=1 files=500 dir='rel' name='/index.js'           ***    -11.99 %       ±5.22%  ±6.96%  ±9.08%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='abs' name='/'                   *     -2.57 %       ±2.17%  ±2.89%  ±3.77%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='abs' name=''                          -1.32 %       ±2.00%  ±2.66%  ±3.47%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='abs' name='/index.js'                 -0.64 %       ±2.79%  ±3.72%  ±4.87%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='rel' name='/'                          1.00 %       ±4.80%  ±6.43%  ±8.47%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='rel' name=''                           1.05 %       ±2.33%  ±3.10%  ±4.05%
11:28:00  module/module-loader.js cache='true' n=1000 files=500 dir='rel' name='/index.js'           *      3.60 %       ±3.51%  ±4.68%  ±6.10%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='abs' name='/'                             2.87 %       ±5.74%  ±7.63%  ±9.94%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='abs' name=''                             -0.71 %       ±5.32%  ±7.09%  ±9.27%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='abs' name='/index.js'                     0.34 %       ±4.99%  ±6.67%  ±8.75%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='rel' name='/'                             2.18 %       ±6.47%  ±8.61% ±11.21%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='rel' name=''                              2.39 %       ±3.03%  ±4.04%  ±5.27%
11:28:00  module/module-loader.js cache='true' n=1 files=500 dir='rel' name='/index.js'                    -4.25 %       ±6.25%  ±8.32% ±10.83%
```

Dragiyski commented:
> From the benchmark CI, this seems to introduce a significant performance regression in the module loader

It seems the performance impact comes from ``pathToFileURL``, that function does a lot of things, some of which might be unnecessary (but that's not easy to validate).

A cache is not a viable solution here. A path cache will make the performance test improves without actually improving performance. The test perform ``require`` on two files and then remove it from ``Module._cache`` and repeat 10000 times, simulating loading of large number of modules (that's ``module/module-loader-circular.js``). A path cache would improve the benchmark, because if there is only 2 paths the ``pathToFileURL`` would be called only 2 times. However, the performance for loading actual 10000 different modules with different paths would remain negatively impacted. ``Module._cache`` actually already does a cache ensuring every module is compiled once.

A possible alternative to that is to first check if ``path.isAbsolute`` and if so, update a single cached ``URL`` object ``pathname`` and get the ``href``. This decreases the performance with 20-40% (it wouldn't be -16%, but -10%, still significant). However ``pathToFileURL`` makes a few checks about the windows ``\\X\`` type of paths and I am not sure if CJS module could be on such a path or not.

Overall, the problem is within ``pathToFileURL``. If that could be optimized, it will reduce the impact. I tried to find why this is not a problem for ESM modules since they have their path normalized by that function. However, it seems this is not a problem, since ``import()`` is ``async`` and ``import`` are handled before any JavaScript code begins execution. Loading huge number of ESM modules with static imports would increase the time required for NodeJS to startup, but once it starts all modules are compiled. And loading huge number of ESM modules with ``import()`` is neither common case, nor it will affect performance, because v8 can compile modules while NodeJS execute metadata.

I am looking for suggestions how to improve that's performance. The ``URL`` methods follows ``WHATWG`` standard and should not modified and the performance impact should affect the application only during early stages of the runtime, because each module is compiled only once.

bcoe commented:
> It seems the performance impact comes from pathToFileURL, that function does a lot of things, some of which might be unnecessary (but that's not easy to validate).

> From the benchmark CI, this seems to introduce a significant performance regression in the module loader

@joyeecheung @Dragiyski perhaps not the ideal, but could we move the resolution logic to the `prepareStackTrace` step? rather than adding additional logic to `loader.js`?

benjamingr reviewed (approved):
Changes look good. I think this should probably land even with the benchmark regression though I believe it's probably a good idea to test this with a real module (like `require('webpack')`) for example and measure the regression there.

Dragiyski committed:
lib: optimize path to file conversion

It seems `wrapSafe` filename is always absolute path. Certain aspects
for validating paths in `pathToFileURL` can be skipped in favor of
better performance of CJS synchrnous loading while keeping the origin.

Dragiyski committed:
fix: lint

Dragiyski commented:
> but could we move the resolution logic to the `prepareStackTrace` step

Yes, this could solve the regression, but I am reluctant to do that, since ``v8::ScriptOrigin`` would remain a path, not a filename, which in case of custom ``Error.prepareStackTrace`` would still result in inconsistent result between ESM and CJS modules (file URL vs path). If no other solution is found, this would do.

In the latest commit I attempted another approach. I stopped using ``pathToFileURL`` due to huge number of checks and errors thrown for certain path in windows, as neither of those checks would be reachable when called from ``wrapSafe``. The path here already passed validation and most likely it is a valid path to existing file. Here is my local benchmark result compared to ``upstream/master``:

```
                                                                                   confidence improvement accuracy (*)   (**)  (***)
 module/module-loader-circular.js n=10000                                                 ***     -6.99 %       ±1.06% ±1.41% ±1.84%
 module/module-loader-deep.js cache='false' files=1000 ext='.js'                          ***     -6.85 %       ±1.64% ±2.18% ±2.83%
 module/module-loader-deep.js cache='false' files=1000 ext=''                             ***     -5.96 %       ±1.87% ±2.49% ±3.24%
 module/module-loader-deep.js cache='true' files=1000 ext='.js'                                    0.66 %       ±3.78% ±5.04% ±6.56%
 module/module-loader-deep.js cache='true' files=1000 ext=''                                      -2.90 %       ±5.17% ±6.87% ±8.95%
 module/module-loader.js cache='false' n=1 files=500 dir='abs' name=''                    ***     -6.32 %       ±2.22% ±2.96% ±3.85%
 module/module-loader.js cache='false' n=1 files=500 dir='abs' name='/'                   ***     -6.25 %       ±2.48% ±3.30% ±4.30%
 module/module-loader.js cache='false' n=1 files=500 dir='abs' name='/index.js'            **     -3.61 %       ±2.40% ±3.21% ±4.19%
 module/module-loader.js cache='false' n=1 files=500 dir='rel' name=''                    ***     -5.93 %       ±1.90% ±2.53% ±3.29%
 module/module-loader.js cache='false' n=1 files=500 dir='rel' name='/'                   ***     -5.50 %       ±2.38% ±3.16% ±4.12%
 module/module-loader.js cache='false' n=1 files=500 dir='rel' name='/index.js'           ***     -6.82 %       ±2.40% ±3.20% ±4.17%
 module/module-loader.js cache='false' n=1000 files=500 dir='abs' name=''                          0.20 %       ±2.08% ±2.76% ±3.59%
 module/module-loader.js cache='false' n=1000 files=500 dir='abs' name='/'                        -1.41 %       ±2.47% ±3.29% ±4.29%
 module/module-loader.js cache='false' n=1000 files=500 dir='abs' name='/index.js'                -0.91 %       ±3.38% ±4.50% ±5.87%
 module/module-loader.js cache='false' n=1000 files=500 dir='rel' name=''                         -0.87 %       ±2.51% ±3.34% ±4.34%
 module/module-loader.js cache='false' n=1000 files=500 dir='rel' name='/'                         0.25 %       ±2.94% ±3.92% ±5.10%
 module/module-loader.js cache='false' n=1000 files=500 dir='rel' name='/index.js'                 0.86 %       ±3.04% ±4.04% ±5.27%
 module/module-loader.js cache='true' n=1 files=500 dir='abs' name=''                             -2.48 %       ±2.89% ±3.84% ±5.00%
 module/module-loader.js cache='true' n=1 files=500 dir='abs' name='/'                             2.15 %       ±3.21% ±4.27% ±5.57%
 module/module-loader.js cache='true' n=1 files=500 dir='abs' name='/index.js'                    -1.27 %       ±3.51% ±4.68% ±6.11%
 module/module-loader.js cache='true' n=1 files=500 dir='rel' name=''                             -1.33 %       ±2.54% ±3.38% ±4.40%
 module/module-loader.js cache='true' n=1 files=500 dir='rel' name='/'                             1.12 %       ±2.21% ±2.95% ±3.85%
 module/module-loader.js cache='true' n=1 files=500 dir='rel' name='/index.js'                     2.48 %       ±4.07% ±5.43% ±7.09%
 module/module-loader.js cache='true' n=1000 files=500 dir='abs' name=''                   **     -3.22 %       ±2.21% ±2.94% ±3.83%
 module/module-loader.js cache='true' n=1000 files=500 dir='abs' name='/'                         -0.21 %       ±2.21% ±2.95% ±3.84%
 module/module-loader.js cache='true' n=1000 files=500 dir='abs' name='/index.js'                 -0.94 %       ±1.68% ±2.24% ±2.91%
 module/module-loader.js cache='true' n=1000 files=500 dir='rel' name=''                          -0.41 %       ±1.69% ±2.25% ±2.93%
 module/module-loader.js cache='true' n=1000 files=500 dir='rel' name='/'                         -0.42 %       ±1.48% ±1.97% ±2.57%
 module/module-loader.js cache='true' n=1000 files=500 dir='rel' name='/index.js'                  0.87 %       ±2.38% ±3.17% ±4.13%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case there are 29 comparisons, you can thus
expect the following amount of false-positive results:
  1.45 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.29 false positives, when considering a   1% risk acceptance (**, ***),
  0.03 false positives, when considering a 0.1% risk acceptance (***)
```

There is still significant regression, but not as severe as before.

Trott commented:
@Dragiyski Is this ready for review/landing in your view, or is there still some work you're hoping to do on it first?

Dragiyski commented:
> @Dragiyski Is this ready for review/landing in your view, or is there still some work you're hoping to do on it first?

Yes, I think it is ready for review.

bcoe commented:
I agree with @benjamingr's recommendation, let's benchmark a couple real world modules that are quite large, pulling in the index of [googleapis](https://www.npmjs.com/package/googleapis) would be a great candidate (_it's dependency graph is huge, and probably represents a good _worst case_).

Ultimately, I'll trust to @joyeecheung and @benjamingr, as to whether the regression is acceptable.

Otherwise, I'm happy with this functionality and am excited that it makes behavior more consistent between ESM and CJS 👍

Trott commented:
@nodejs/tsc (Semver major so needs at least 2 TSC approvals.)

joyeecheung commented:
I am still not sure whether making inspector stack traces nicer (a mostly dev-only feature) is worth the cost of the performance regression and potential breakage in the ecosystem.  I think we should at least investigate how this would impact real-world packages.

bcoe commented:
> Yes, this could solve the regression, but I am reluctant to do that, since v8::ScriptOrigin would remain a path, not an URL, which in case of custom Error.prepareStackTrace would still result in inconsistent result between ESM and CJS modules (file URL vs path). If no other solution is found, this would do.

@joyeecheung @Dragiyski why don't we start with improvement in `Error.prepareStackTrace`, since it's less controversial?

joyeecheung commented:
@bcoe Starting with the `Error.prepareStackTrace` sounds good, since we are likely to be dealing with errors in those cases

aduh95 commented:
@Dragiyski can you please rebase on top of master to solve the git conflict?

bcoe commented:
@Dragiyski please let me know when you have something that you'd like folks to take another look at.

bcoe commented:
@Dragiyski please let me know when you have something that you'd like folks to take another look at.

aduh95 commented:
This needs a rebase.

aduh95 commented:
@Dragiyski would be able to rebase on top of `master` to solve the git conflict please?
